pallets/idn-consumer/src/lib.rs

```diff
@@ -0,0 +1,231 @@
+		///
+		/// **Example definition**
+		/// ```nocompile
+		/// pub SiblingIdnParaId: u32 = IDN_PARACHAIN_ID;
+		/// pub SiblingIdnLocation: Location = Location::new(1, Parachain(SiblingIDNParaId::get()));
+		/// ```
+		type SiblingIdnLocation: Get<Location>;
+
+		/// The origin of the IDN chain
+		///
+		/// **Example definition**
+		/// ```nocompile
+		/// pub IdnOrigin: EnsureXcm<Equals<Self::SiblingIdnLocation>>
+		/// ```
+		type IdnOrigin: EnsureOrigin<Self::RuntimeOrigin, Success = Location>;
+
+		/// A type to define the amount of credits in a subscription
+		type Credits: Unsigned + Encode;
+
+		/// Maximum metadata size
+		type MaxMetadataLen: Get<u32>;
+
+		/// Subscription ID type
+		type SubscriptionId: From<H256>
+			+ Codec
+			+ Copy
+			+ PartialEq
+			+ TypeInfo
+			+ EncodeLike
+			+ MaxEncodedLen
+			+ Debug;
+
+		/// Maximum Pulse Filter size
+		type MaxPulseFilterLen: Get<u32>;
+	}
+
+	#[pallet::pallet]
+	pub struct Pallet<T>(_);
+
+	#[pallet::error]
+	pub enum Error<T> {
+		/// An error occurred while consuming the pulse
+		ConsumeError,
+	}
+
+	#[pallet::event]
+	#[pallet::generate_deposit(pub(super) fn deposit_event)]
+	pub enum Event<T: Config> {
+		/// A random value was successfully consumed.
+		RandomnessConsumed { round: <T::Pulse as Pulse>::Round, sub_id: T::SubscriptionId },
+	}
+
+	#[pallet::call]
+	impl<T: Config> Pallet<T> {
+		/// Creates a subscription.
+		#[pallet::call_index(0)]
+		#[pallet::weight(Weight::from_parts(0, 0))]
+		#[allow(clippy::useless_conversion)]
+		pub fn consume(
+			origin: OriginFor<T>,
+			pulse: T::Pulse,
+			sub_id: T::SubscriptionId,
+		) -> DispatchResultWithPostInfo {
+			// ensure origin is coming from IDN
+			let _ = T::IdnOrigin::ensure_origin(origin)?;
+
+			let round = pulse.round().clone();
+
+			T::Consumer::consume(pulse, sub_id)?;
+
+			Self::deposit_event(Event::RandomnessConsumed { round, sub_id });
+
+			Ok(Pays::No.into())
+		}
+	}
+}
+
+impl<T: Config> Pallet<T> {
+	/// Creates a subscription.
+	pub fn create_subscription(
+		// Number of random values to receive
+		credits: T::Credits,
+		// Distribution interval for pulses
+		frequency: BlockNumberFor<T>,
+		// Bounded vector for additional data
+		metadata: Option<MetadataOf<T>>,
+		// Optional Pulse Filter
+		pulse_filter: Option<PulseFilterOf<T>>,
+		// Optional Subscription Id, if None, a new one will be generated
+		sub_id: Option<T::SubscriptionId>,
+	) -> Result<T::SubscriptionId, Error<T>> {
+		// TODO: finish implementation https://github.com/ideal-lab5/idn-sdk/issues/81
+
+		let mut params = CreateSubParamsOf::<T> {
+			credits,
+			target: T::SiblingIdnLocation::get(),
+			call_index: [0, 0],
+			frequency,
+			metadata,
+			pulse_filter,
+			sub_id,
+		};
+
+		// If sub_id is not provided, generate a new one and asign it to the params
+		let sub_id = match sub_id {
+			Some(sub_id) => sub_id,
+			None => {
+				let salt = frame_system::Pallet::<T>::block_number().encode();
```

Comment:
This would result in a collision if two origins submit transactions to create subscriptions with the same parameters in the same block - it should be good enough if we concat the origin (so, adding a new parameter to this function) to the hasher, so instead of having Hash(params || block_number), we have Hash(params || origin_pk || block_number).